phase/upgrade_controllers.go

```diff
@@ -0,0 +1,109 @@
+package phase
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/Masterminds/semver"
+	"github.com/k0sproject/k0sctl/config"
+	"github.com/k0sproject/k0sctl/config/cluster"
+	log "github.com/sirupsen/logrus"
+)
+
+// UpgradeControllers upgrades the controllers one-by-one
+type UpgradeControllers struct {
+	GenericPhase
+
+	hosts cluster.Hosts
+}
+
+// Title for the phase
+func (p *UpgradeControllers) Title() string {
+	return "Upgrade controllers"
+}
+
+// Prepare the phase
+func (p *UpgradeControllers) Prepare(config *config.Cluster) error {
+	log.Debugf("UpgradeControllers phase prep starting")
+	p.Config = config
+	var controllers cluster.Hosts = p.Config.Spec.Hosts.Controllers()
+	log.Debugf("%d controllers in total", len(controllers))
+	p.hosts = controllers.Filter(func(h *cluster.Host) bool {
+		return h.Metadata.NeedsUpgrade
+	})
+	log.Debugf("UpgradeControllers phase prepared, %d controllers needs upgrade", len(p.hosts))
+	return nil
+}
+
+// ShouldRun is true when there are controllers that needs to be upgraded
+func (p *UpgradeControllers) ShouldRun() bool {
+	return len(p.hosts) > 0
+}
+
+// Run the phase
+func (p *UpgradeControllers) Run() error {
+	for _, h := range p.hosts {
+		log.Infof("%s: starting upgrade", h)
+		if p.needsMigration(h) {
+			if err := p.migrateService(h); err != nil {
+				return err
+			}
+		} else {
+			if err := h.Configurer.StopService(h, h.K0sServiceName()); err != nil {
+				return err
+			}
+		}
+		if err := h.UpdateK0sBinary(p.Config.Spec.K0s.Version); err != nil {
+			return err
+		}
+		if err := h.Configurer.StartService(h, h.K0sServiceName()); err != nil {
+			return err
+		}
+		log.Infof("%s: waiting for the k0s service to start", h)
+		if err := h.WaitK0sServiceRunning(); err != nil {
+			return err
+		}
+		if err := h.WaitKubeAPIReady(); err != nil {
+			return err
+		}
+
+	}
+	return nil
+}
+
+func (p *UpgradeControllers) needsMigration(h *cluster.Host) bool {
+	log.Debugf("%s: checking need for 0.10 --> 0.11 migration", h)
+	c, _ := semver.NewConstraint("< 0.11")
+	current, err := semver.NewVersion(h.Metadata.K0sRunningVersion)
+	if err != nil {
+		log.Warnf("%s: failed to parse version info: %s", h, err.Error())
+		return false
+	}
+
+	return c.Check(current)
```

Comment:
There's something weird in the semver package.

https://play.golang.org/p/iGEGmNWEvdC

```
0.10.0-rc2 < 0.11 check result: false
```

Reply:
> SemVer comparisons using constraints without a prerelease comparator will skip prerelease versions. For example, >=1.2.3 will skip prereleases when looking at a list of releases while >=1.2.3-0 will evaluate and find prereleases.

So making the constraint as `< 0.11-0` "fixes" it

Reply:
@kke this is now fixed in https://github.com/k0sproject/k0sctl/pull/93/commits/43f02e26aa6cae3f9704203eb6754e4e680bb157